### PR TITLE
Date picker week numbers settings

### PR DESCRIPTION
**Pull request by yananym**
#### Pull request checklist

- [-] Addresses an existing issue: 
- [+] Include a change request file using `$ npm run change`

#### Description of changes

Adding new feature to calendar inside of DatePicket to calculate week numbers based on settings passed to a date picker (start counting weeks on Jan 1st/first full week or first 4-day week.
@johannao76 @kellygorr 

#### Focus areas to test

DatePicker with week numbers shown: e.g. four-day-week setting is enabled on screenshot:
screen shot 2017-10-02 at 1 41 17 pm

### CODE REVIEW

**Comment by micahgodbolt** [comment]
@yananym you've got a bunch of linting errors due to using double, instead of single quotes.

File: packages/office-ui-fabric-react/src/components/Calendar/Calendar.tsx [hunk 1]
```diff
@@ -72,15 +88,22 @@ export class Calendar extends BaseComponent<ICalendarProps, ICalendarState> impl
 
   constructor(props: ICalendarProps) {
     super(props);
-    let currentDate = props.value && !isNaN(props.value.getTime()) ? props.value : (props.today || new Date());
+    let currentDate =
+      props.value && !isNaN(props.value.getTime())
```
Comment (kellygorr) [inline packages/office-ui-fabric-react/src/components/Calendar/Calendar.tsx]: It looks like new formatting was applied to many areas of the files (even though you did not edit them)!

**Comment by kellygorr** [comment]
When firstWeekOfYear={ 2 } (first 4-day week of year) The last week of December should be week 1 instead of 52.  When year is 2000, the first week of the year is also off (although this issue - 54 weeks in a year- only appears to happen every 28 years?)

Screenshot from current owa calendar (first 4-day week of year)
test1

test2

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts [hunk 2]
```diff
@@ -207,11 +205,78 @@ function getDatePartHashValue(date: Date) {
  * @param {DayOfWeek} firstDayOfWeek - The first day of the week (0-6, Sunday = 0)
  * @return {Number} The week's number in the year.
  */
-export function getWeekNumber(date: Date, firstDayOfWeek: DayOfWeek) {
-  let num = getDayOfYear(date) - 1;
-  let num2 = date.getDay() - (num % 7);
-  let num3 = ((num2 - firstDayOfWeek) + 14) % 7;
-  return Math.floor(((num + num3) / 7) + 1);
+export function getWeekNumber(date: Date, firstDayOfWeek: DayOfWeek, firstWeekOfYear: FirstWeekOfYear) {
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: Kelly had already added code for week numbers. Are you adding your own also? We should have one way to do week numbers. Kelly was using the method I recommended based on JSMVVM implementation so I suggest we stick with that and just augment with the firstWeekOfYear setting support

File: packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts [hunk 3]
```diff
@@ -12,10 +12,19 @@ export enum DayOfWeek {
 }
 
 /**
+ * First week of the year settings types
+ */
+export enum FirstWeekOfYear {
+  FirstDay = 0,
+  FirstFullWeek = 1,
+  FirstFourDayWeek = 2
+}
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts]: I assume this maps to the JSMVVM values, correct? Once you commit these changes and we upgrade owa-calendar to use the new version, we should remove the enum from owa-calendar and use this one.

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts (hunk 2 above)
Comment (yananym) [reply packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: That's exactly what i'm doing. Her function is getDefaultWeekNumber() - I renames variables to make a little more sense of the math.

File: packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts (hunk 3 above)
Comment (yananym) [reply packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts]: yes it is what being passed through settings.

File: packages/office-ui-fabric-react/src/utilities/dateValues/TimeConstants.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  MillisecondsInOneDay: 86400000,
+  MillisecondsIn1Sec: 1000,
+  MillisecondsIn1Min: 60000,
+  MillisecondsIn30Mins: 1800000,
+  MillisecondsIn1Hour: 3600000,
+  MinutesInOneDay: 1440,
+  MinutesInOneHour: 60,
+  DaysInOneWeek: 7,
+  MonthInOneYear: 12,
+  WeeksInOneYear: 52
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateValues/TimeConstants.ts]: Expose only the things you need, are you using all these? Also, OWA JSMVVM shows 53 weeks for this year:
image

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts (hunk 2 above)
Comment (johannao76) [reply packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: Also, we should add unit tests

File: packages/office-ui-fabric-react/src/utilities/dateValues/TimeConstants.ts (hunk 4 above)
Comment (yananym) [reply packages/office-ui-fabric-react/src/utilities/dateValues/TimeConstants.ts]: @johannao76 what do you have set at firstWeekOFYear?

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts [hunk 5]
```diff
@@ -205,13 +203,92 @@ function getDatePartHashValue(date: Date) {
...
+* Helper function for getWeekNumber.
+* Returns adjusted week day number when firstDayOfWeek is other than Sunday
+* For Week Day Number comparison checks
+* @param {firstDayOfWeek} DayOfWeek - The first day of week (0-6, Sunday = 0)
+* @param {dateWeekDay} DayOfWeek - shifts number forward to 1 week in case passed as true
+* @return {DayOfWeek} The day of week adjusted to firstDayOfWeek. E.g. when FirstDyOfWeek is Monday (1), Sunday becomes = 7 (7 > 1).
+*/
+export function adjustWeekDay(firstDayOfWeek: DayOfWeek, dateWeekDay: DayOfWeek) {
+  return (firstDayOfWeek !== DayOfWeek.Sunday && dateWeekDay < firstDayOfWeek) ? dateWeekDay + TimeConstants.DaysInOneWeek : dateWeekDay;
 }
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: I am still not seeing unit tests for this logic, please add unit tests

**Comment by johannao76** [comment]
Bug: December 2018 shows two week numbers in the new implementation depending on which month you are looking at:
image
image
In JSMVVM, it always shows it correctly as week 1:
image
image

**Comment by johannao76** [comment]
Please add the unit tests for the different cases, particularly around boundary conditions which we know are bound to cause problems.

File: packages/office-ui-fabric-react/src/components/Calendar/Calendar.tsx [hunk 6]
```diff
@@ -58,6 +58,7 @@ export class Calendar extends BaseComponent<ICalendarProps, ICalendarState> impl
     highlightCurrentMonth: false,
     navigationIcons: iconStrings,
     showWeekNumbers: false,
+    firstWeekOfYear: 0,
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/components/Calendar/Calendar.tsx]: >0 [](start = 21, length = 1)

FirstWeekOfYear.FirstDay

File: packages/office-ui-fabric-react/src/components/Calendar/CalendarDay.tsx [hunk 7]
```diff
@@ -495,18 +496,25 @@ export class CalendarDay extends BaseComponent<ICalendarDayProps, ICalendarDaySt
   }
 
   // Returns the week numbers for each week in a month.  Week numbers are 1 - 52 (53) in a year
-  private _getWeekNumbersInMonth(weeksInMonth: number, firstDayOfWeek: DayOfWeek, navigatedDate: Date) {
+  private _getWeekNumbersInMonth(weeksInMonth: number, firstDayOfWeek: DayOfWeek, firstWeekOfYear: FirstWeekOfYear, navigatedDate: Date) {
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/components/Calendar/CalendarDay.tsx]: >_getWeekNumbersInMonth [](start = 10, length = 22)

Do we need to move getWeekNumbersInMonth() to DateMath also so we can call it from owa-calendar?

Comment (yananym) [reply packages/office-ui-fabric-react/src/components/Calendar/CalendarDay.tsx]: good thought!

File: packages/office-ui-fabric-react/src/components/DatePicker/DatePicker.Props.ts [hunk 8]
```diff
@@ -117,6 +118,12 @@ export interface IDatePickerProps extends React.Props<DatePicker> {
   showWeekNumbers?: boolean;
 
   /**
+  * Defines when the first week of the year should start, 0, 1 or 2 are the possible values
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/components/DatePicker/DatePicker.Props.ts]: > 0, 1 or 2 [](start = 57, length = 10)

Let's be more clear about what 0, 1, and 2 are. This is a public API, we should indicate that the possible values are: FirstWeekOfYear.FirstDay, FirstWeekOfYear.FirstFullWeek, etc

File: packages/office-ui-fabric-react/src/components/DatePicker/DatePicker.tsx [hunk 9]
```diff
@@ -114,6 +115,7 @@ export class DatePicker extends BaseComponent<IDatePickerProps, IDatePickerState
     borderless: false,
     pickerAriaLabel: 'Calender',
     showWeekNumbers: false,
+    firstWeekOfYear: 0,
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/components/DatePicker/DatePicker.tsx]: >0 [](start = 21, length = 1)

FirstWeekOfYear.FirstDay

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts [hunk 10]
```diff
@@ -292,4 +292,52 @@ describe('DateMath', () => {
     dateRangeArray = DateMath.getDateRangeArray(date, DateRangeType.Week, DayOfWeek.Tuesday);
     Array(7).forEach((val, i) => expect(DateMath.compareDates(dateRangeArray[i], date)).to.be.true);
   });
+
+  it('can calculate week numbers - option 0', () => {
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts]: >option 0 [](start = 35, length = 8)

with first week of year set to first day of year

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts [hunk 11]
```diff
@@ -292,4 +292,52 @@ describe('DateMath', () => {
...
+    expect(DateMath.getWeekNumber(date1, 0, 0) === 1).to.be.true;
+
+    date1 = new Date(2019, 0, 1);
+    expect(DateMath.getWeekNumber(date1, 0, 0) === 1).to.be.true;
+
+    date1 = new Date(2010, 11, 31);
+    expect(DateMath.getWeekNumber(date1, 1, 0) === 53).to.be.true;
+  });
+
+  it('can calculate week numbers - option 1', () => {
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts]: >option 1 [](start = 35, length = 8)

with first week of year set to xyz, you get the idea :)

**Comment by johannao76** [comment]
Can you add an example card for a different first day of the week value?

File: packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts (hunk 3 above)
Comment (johannao76) [reply packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts]: make sure you export per our discussion offline so we can reuse in owa-calendar

---
In reply to: [142754811](https://github.com/OfficeDev/office-ui-fabric-react/pull/3001#discussion_r142754811) [](ancestors = 142754811)

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts [hunk 12]
```diff
@@ -23,272 +24,320 @@ describe('DateMath', () => {
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts]: >.to.be.true; [](start = 53, length = 12)

mentione offline, this needs to be .toBe(true);

Comment (yananym) [reply packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts]: have it as toEqual()

**Comment by johannao76** [comment]
Can you please open a bug for that issue I sent you around firstDayOfWeek = firstFourDayWeek setting showing dupe week numbers in Dec 2020?

File: yarn.lock [hunk 13]
```diff
@@ -0,0 +1,1817 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
```
Comment (johannao76) [inline yarn.lock]: remove this file?

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts [hunk 14]
```diff
@@ -200,18 +198,119 @@ function getDatePartHashValue(date: Date) {
   /* tslint:enable:no-bitwise */
 }
 
+// Returns the week numbers for each week in a month.  Week numbers are 1 - 52 (53) in a year
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: Add the exports before the private functions and follow the existing commenting conventions please

Comment (yananym) [reply packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.ts]: #Resolved

File: packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts (hunk 3 above)
Comment (yananym) [reply packages/office-ui-fabric-react/src/utilities/dateValues/DateValues.ts]: #Resolved

File: packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts [hunk 15]
```diff
@@ -291,4 +291,97 @@ describe('DateMath', () => {
     dateRangeArray = DateMath.getDateRangeArray(date, DateRangeType.Week, DayOfWeek.Tuesday);
     Array(7).forEach((val, i) => expect(DateMath.compareDates(dateRangeArray[i], date)).toBe(true));
   });
+
+  // First week of year set to FirstWeekOfYear.FirstDay
```
Comment (johannao76) [inline packages/office-ui-fabric-react/src/utilities/dateMath/DateMath.test.ts]: >First [](start = 5, length = 5)

Need a few tests for getWeekNumbersInMonth

**Comment by yananym** [comment]
Bug https://github.com/OfficeDev/office-ui-fabric-react/issues/3119

**Comment by johannao76** [comment]
Thanks! Also opened a corresponding one in our VSO just so we dont forget: [https://outlookweb.visualstudio.com/Outlook%20Web/_workitems?id=20427&_a=edit](https://outlookweb.visualstudio.com/Outlook%2520Web/_workitems?id=20427&_a=edit)

---
In reply to: [336580093](https://github.com/OfficeDev/office-ui-fabric-react/pull/3001#issuecomment-336580093) [](ancestors = 336580093)
